### PR TITLE
Runtime Interface Reflection: rmw

### PR DESCRIPTION
**Pull request by methylDragon**
This PR is part of the runtime interface reflection subscription feature of REP-2011: https://github.com/ros2/ros2/issues/1374

# Description
This PR includes changes to the RMW APIs for supporting runtime interface reflection, namely.
- Introducing a new rmw agnostic typesupport `rosidl_message_type_support_t`

With also new interfaces for:
- Taking and handling runtime type messages (dynamic data)
- Using the serialization support library

### CODE REVIEW

File: rmw/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 endif()
 
 find_package(ament_cmake_ros REQUIRED)
+find_package(rosidl_dynamic_typesupport REQUIRED)
```
Comment (wjwwood) [inline rmw/CMakeLists.txt:21]: move below with other deps

File: rmw/CMakeLists.txt [hunk 2]
```diff
@@ -75,6 +80,7 @@ ament_export_libraries(${PROJECT_NAME})
 
 # Export modern CMake targets
 ament_export_targets(${PROJECT_NAME})
+ament_export_dependencies(rosidl_dynamic_typesupport)
```
Comment (wjwwood) [inline rmw/CMakeLists.txt:83]: move up with other export deps

File: rmw/package.xml [hunk 3]
```diff
@@ -16,10 +16,12 @@
...
 
+  <build_depend>rosidl_dynamic_typesupport</build_depend>
   <build_depend>rcutils</build_depend>
   <!-- Only needed because CMake versions less than 3.13 don't support CMP0079 -->
   <build_depend>rosidl_runtime_c</build_depend>
 
+  <build_export_depend>rosidl_dynamic_typesupport</build_export_depend>
   <build_export_depend>rcutils</build_export_depend>
   <!-- This is required for the definition of the rosidl typesupport types -->
   <build_export_depend>rosidl_runtime_c</build_export_depend>
```
Comment (wjwwood) [inline rmw/package.xml:27]: alphabetize

Comment (methylDragon) [reply rmw/package.xml]: https://github.com/ros2/rmw/pull/340/commits/af30f1d55cfb68fee77aeab7d5ec78c46b593648

File: rmw/CMakeLists.txt (hunk 2 above)
Comment (methylDragon) [reply rmw/CMakeLists.txt]: https://github.com/ros2/rmw/pull/340/commits/af30f1d55cfb68fee77aeab7d5ec78c46b593648

File: rmw/CMakeLists.txt (hunk 1 above)
Comment (methylDragon) [reply rmw/CMakeLists.txt]: https://github.com/ros2/rmw/pull/340/commits/af30f1d55cfb68fee77aeab7d5ec78c46b593648

File: rmw/src/dynamic_message_type_support.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  if (success) {
+    return RMW_RET_OK;
+  } else {
+    RMW_SET_ERROR_MSG(
+      "could not deserialize serialized message to dynamic data: "
+      "dynamic data not enough memory");
+    return ret;
+  }
+}
```
Comment (wjwwood) [inline rmw/src/dynamic_message_type_support.c:134]: I think this one should also be deleted, as it doesn't seem to provide any value over calling the `rosidl_dynamic_typesupport` version of this. I was going to comment that `success` doesn't necessary imply there wasn't enough memory, or if it does, then that should just be a different return code, but I think we just remove this entirely from `rmw`.

File: rmw/src/dynamic_message_type_support.c [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    rosidl_dynamic_typesupport_dynamic_data_serialize(
+      dynamic_message, serialized_message, &success));
+
+  if (success) {
+    return RMW_RET_OK;
+  } else {
+    RMW_SET_ERROR_MSG("could not serialize dynamic data to serialized message!");
+    return ret;
+  }
+}
```
Comment (wjwwood) [inline rmw/src/dynamic_message_type_support.c:152]: Same for this one, probably should just remove it.

File: rmw/src/dynamic_message_type_support.c [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(ts, RMW_RET_INVALID_ARGUMENT);
+  return rmw_convert_rcutils_ret_to_rmw_ret(rosidl_dynamic_message_type_support_handle_destroy(ts));
+}
+
+const char *
+rmw_get_dynamic_typesupport_identifier(void)
+{
+  return rosidl_dynamic_typesupport_c__identifier;
+}
```
Comment (wjwwood) [inline rmw/src/dynamic_message_type_support.c:84]: Also, I think we should move this function to `rosidl_dynamic_typesupport`.

**Review by wjwwood — CHANGES_REQUESTED** [review]
We have some interfaces to remove/move, but otherwise lgtm

File: rmw/src/dynamic_message_type_support.c (hunk 6 above)
Comment (methylDragon) [reply rmw/src/dynamic_message_type_support.c]: Moved! (See rosidl_dynamic_typesupport PR https://github.com/ros2/rosidl_dynamic_typesupport/pull/1)

File: rmw/src/dynamic_message_type_support.c (hunk 5 above)
Comment (methylDragon) [reply rmw/src/dynamic_message_type_support.c]: https://github.com/ros2/rmw/pull/340/commits/4de8ad65e91bc088af0a4a456e9a179bc348f5bb

File: rmw/src/dynamic_message_type_support.c (hunk 4 above)
Comment (methylDragon) [reply rmw/src/dynamic_message_type_support.c]: https://github.com/ros2/rmw/pull/340/commits/4de8ad65e91bc088af0a4a456e9a179bc348f5bb

File: rmw/include/rmw/dynamic_message_type_support.h [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+rmw_create_dynamic_message_from_dynamic_message_type(
+  rosidl_dynamic_typesupport_dynamic_type_t * dynamic_message_type,
+  rosidl_dynamic_typesupport_dynamic_data_t ** cloned_dynamic_message_type);  // OUT
+
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_ret_t
+rmw_clone_dynamic_message(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_message,
+  rosidl_dynamic_typesupport_dynamic_data_t ** cloned_dynamic_message);  // OUT
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply rmw/include/rmw/dynamic_message_type_support.h]: https://github.com/ros2/rmw/pull/340/commits/4de8ad65e91bc088af0a4a456e9a179bc348f5bb

**Review by wjwwood — APPROVED** [review]

**Pull request merged**